language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+
+use anyhow::{bail, format_err, Result};
+use diem_client::{
+    views::AccountStateWithProofView, AccountAddress, BlockingClient, MethodRequest, MethodResponse,
+};
+use diem_read_write_set::ReadWriteSetAnalysis;
+use diem_state_view::StateView;
+use diem_transaction_replay::DiemDebugger;
+use diem_types::access_path::AccessPath;
+use diem_types::transaction::{TransactionOutput, Version};
+use diem_types::{
+    account_state::AccountState, account_state_blob::AccountStateBlob, transaction::Transaction,
+};
+use diem_validator_interface::{DiemValidatorInterface, JsonRpcDebuggerInterface};
+use diem_vm::VMExecutor;
+use move_cli::sandbox::utils::{
+    mode::{Mode, ModeType},
+    on_disk_state_view::OnDiskStateView,
+};
+use move_core_types::language_storage::ResourceKey;
+
+use structopt::StructOpt;
+
+pub struct OnDiskStateViewReader {
+    view: OnDiskStateView,
+}
+
+struct AccountAddressWithState {
+    address: AccountAddress,
+    state: AccountState,
+}
+
+#[derive(Debug)]
+pub struct PrefetchTransactionReplayerConfig {
+    build_dir: PathBuf,
+    storage_dir: PathBuf,
+    rpc_url: String,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    about = "Replay transactions using state prefetching",
+    rename_all = "kebab-case"
+)]
+pub struct PrefetchedReplayCLI {
+    #[structopt(long, default_value = "http://localhost:8080")]
+    rpc_url: String,
+
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+#[derive(Debug, StructOpt)]
+pub enum Command {
+    /// Replay transactions starting from version `start` to `start + limit`.
+    #[structopt(name = "replay-transactions")]
+    ReplayTransactions { start: Version, limit: u64 },
+}
+
+pub struct PrefetchTransactionReplayer {
+    config: PrefetchTransactionReplayerConfig,
+    validator: Box<dyn DiemValidatorInterface>,
+    debugger: DiemDebugger,
+    rpc_client: BlockingClient,
+}
+
+impl StateView for OnDiskStateViewReader {
+    fn get(&self, access_path: &AccessPath) -> Result<Option<Vec<u8>>> {
+        access_path.get_struct_tag().map_or(Ok(None), |tag| {
+            self.view.get_resource_bytes(access_path.address, tag)
+        })
+    }
+
+    fn is_genesis(&self) -> bool {
+        false
+    }
+}
+
+fn ensure_dir_does_not_exist(dir: &Path) -> Result<()> {
+    if dir.exists() {
+        Err(format_err!("{:?} must be empty as it will be created", dir))
+    } else {
+        Ok(())
+    }
+}
+
+impl PrefetchTransactionReplayer {
+    pub fn from_config(config: PrefetchTransactionReplayerConfig) -> Result<Self> {
+        ensure_dir_does_not_exist(&config.build_dir)?;
+        ensure_dir_does_not_exist(&config.storage_dir)?;
+
+        let debugger = DiemDebugger::json_rpc_with_config(
+            &config.rpc_url,
+            config.build_dir.clone(),
+            config.storage_dir.clone(),
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        // NOTE: we could probably avoid creating a second validator
```

Comment:
I think that is also a reasonable change to make inside `DiemDebugger` (CC @runtian-zhou)